### DISCUSSION TITLE
Use React DayPicker component with use-local-storage-state hook

### DISCUSSION THREAD

**Discussion by AliaksandraNovikava** [neuefische/questions#645] [Web] [answered]
Hi,
In my project I'm using "DayPicker" (a date picker component for React) that renders a monthly calendar to select days. To be more specific, I'm using this one: [Selecting Multiple Days](https://react-day-picker.js.org/basics/selecting-days#selecting-multiple-days).
All selected days are saved in a state:
```
  const initialDays = [];
  const [days, setDays] = useState(initialDays);
```
To use this data in an other component I tried to save it in the useLocalStorageState. 
```
  const initialDays = [];
  const [days, setDays] = useLocalStorageState('markedDays', initialDays);
```
Then I got this error:
> Starting with v2.0.0-beta.1 date-fns doesn't accept strings as date arguments. Please use `parseISO` to parse strings.

Then I added this handler function and applied it on the DayPicker component:
```
  const handleDayClick = (day) => {
    const formattedDay = day.toISOString();
    setDays((prevDays) => [...prevDays, formattedDay]);
  };
```
But somehow I could't mark days in the calendar anymore.
Saving days using the useLocalStorageState didn't seem to work so I tried to save the data in a different state using the useLocalStorageState:
```
const [markedDays, setMarkedDays] = useLocalStorageState("markedDays", null);
const handleDayClick = (day) => {
    const clickedDay = day;
    const newMarkedDay = {
      goalId: selectedGoal.id,
      date: clickedDay,
    };
    setMarkedDays((prevMarkedDays) => {
      if (!prevMarkedDays) {
        return [newMarkedDay];
      }
      return [...prevMarkedDays, newMarkedDay];
    });
  };
```
Now all marked days are saved in `markedDays` using useLocalStorageState.
But now I have the problem that every click is saved, even if I click on the same day several times. 
On the other hand, in `days` only days that are actually selected are saved.

After some time I get this log in my console:
```
markedDays: (37) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]
days: (2) ['2023-07-05T22:00:00.000Z', '2023-07-03T22:00:00.000Z']
```

It would be so much simpler if I could somehow save the `days` data in useLocalStorageState.

There is this other problem, that when I mark a day on one goal, the same marked day appears on every other calendar.
But it doesn't affect the saved data. Only days that are actually selected/clicked are saved.

This is my current status: [my app](https://capstone-alex-aim-high-git-feature-e-9c45ce-aliaksandranovikava.vercel.app/). First you will need to set a goal AND repetition to see it on the "My goals" page, then click on the goal to see the details view with the calendar.

UPDATE:
So I tried different approaches to compare the results.
  ```
  function isSameDay(dateLeft: Date | number, dateRight: Date | number): boolean
  namespace isSameDay {}

  const initialDays = [];
  const [days, setDays] = useState(initialDays);
  const [selectedDays, setSelectedDays] = useLocalStorageState("selectedDays", []);
  const [markedDays, setMarkedDays] = useLocalStorageState("markedDays", []);
  
  const myDays = [...days];
  setSelectedDays(myDays);
  
  const handleDayClick = (day) => {
    const clickedDay = day;
    if (!markedDays) {
      setMarkedDays([]);
    }
    const isAlreadyMarked = markedDays.some((markedDay) =>
      isSameDay(markedDay.date, clickedDay)
    );
    if (!isAlreadyMarked) {
      const newMarkedDay = {
        goalId: selectedGoal.id,
        date: clickedDay,
      };
      setMarkedDays((prevMarkedDays) => {
        if (!prevMarkedDays) {
          return [newMarkedDay];
        }
        return [...prevMarkedDays, newMarkedDay];
      });
      setSelectedDays((prevSelectedDays) => {
        if (!prevSelectedDays) {
          return [newMarkedDay];
        }
        return [...prevSelectedDays, newMarkedDay];
      });
    }
  };
```

markedDays: data increases on every click (as expected), but is saved in useLocalStorageState
days, selectedDays: saves only the data of selected days, but gets lost when reloaded

**I need the data from days to be stored in my local storage, either in another state or in the best case there is a way to use the use-local-storage-state hook together with the DayPicker component. So I have for each marked goal a goalId and a date saved in local storage.**

I am grateful for your help and any kind of suggestions🙏

**Comment by actyvystom** [comment]
Hi @AliaksandraNovikava,
as you need to do custom parsing for the state variables passed to `React-Day-Picker`, `useLocalStorageState` won't work here appropriately, so you need to use a more generic way to read and write to localStorage. 

One possible way is to set up a `useEffect` hook having the modal toggle flag as a dependency:

```js
useEffect(() => {
    if (isModalOpen) {
      const dates = JSON.parse(localStorage.getItem("days"));
      if (dates?.length > 0) {
        const transformedDays = dates.map((date) => parseISO(date));
        setDays(transformedDays);
      }
    } else {
      const datesStrings = days.map((day) => day.toISOString());
      localStorage.setItem("days", JSON.stringify(datesStrings));
    }

    return () => {
      const datesStrings = days.map((day) => day.toISOString());
      localStorage.setItem("days", JSON.stringify(datesStrings));
    };
  }, [isModalOpen]);

```
Once the modal is closed, your days data gets written to localstorage in an adequate way. On opening the modal, the data gets read from localstorage and converted to the format needed for the day picker library.

You definitely need to clean up your component and simplify it. Try to separate and unravel the components to have a clean structure. As a suggestion some rough pseudo-code:

```js
<>
   <GoalContentCard/>
   {isModalOpen && (<Modal><DayPicker/></Modal>)}
</>
```

Besides that, you don't need to write the modal toggle flag into localStorage, just keep it as a local state variable. 
I also noticed some wrong usage with `useLocalStorageState` hook when declaring default values. Keep in mind that they are setup in an option object as second argument for the hook call: 

```js
const [markedDays, setMarkedDays] = useLocalStorageState("markedDays", {defaultValue: []});
```

**Reply by AliaksandraNovikava** [reply]
Hi @actyvystom.
Thanks for your reply🙏 It was very helpful, and I made significant progress with the information you provided.
My updated code can be found here: [DayPickerCalendar component](https://github.com/AliaksandraNovikava/capstone-alex/blob/main/components/DayPickerCalendar/index.js)

First I've simplified my `NewGoalDetails` component like you suggested and outsourced code into additional components `Modal` and `DayPickerCalendar` to have a cleaner structure.

And regarding my DayPicker issue:
I had to adjust your code, because the data got lost as soon as I closed the modal. So my current code looks like this:
```
useEffect(() => {
    if (isModalOpen) {
      const dates = JSON.parse(localStorage.getItem("days"));
      if (dates?.length > 0) {
        const transformedDays = dates.map((date) => parseISO(date));
        setDays(transformedDays);
      }
    }
  }, [isModalOpen]);

  useEffect(() => {
    const datesStrings = days.map((day) => day.toISOString());
    localStorage.setItem("days", JSON.stringify(datesStrings));
    const markedDaysObjects = days.map((day) => ({
      goalId: selectedGoal.id,
      date: day,
    }));
    setMarkedDays(markedDaysObjects);
  }, [days, selectedGoal]);
```
I had to adjust the code a little so that I have an object with the goals id and the selected date.

It kind of worked but it created an other problem: every time I switch goals and select a new date all ids in my array are changed to the currently selected goal id. I tried to prevent this behavior using this code:
```
    setMarkedDays((prevMarkedDays) => {
      if (!prevMarkedDays) {
        return markedDaysObjects;
      }
      return [...prevMarkedDays, markedDaysObjects];
    });
```
But it didn't work.

I have been trying to find a solution so that the ids don't change every time I switch a goal for a long time now, got support from fellow students, but I still can't figure out how to solve this problem.

**Reply by actyvystom** [reply]
Hi @AliaksandraNovikava, I think this is probably due to the complex (and somehow confusing) data structure you are using (and also writing to your local storage): 
<img width="460" alt="Bildschirmfoto 2023-07-27 um 15 12 00" src="https://github.com/neuefische/questions/assets/57633320/321a976f-251b-4cc9-a2b3-e77429601f09">
There is `selectedGoal`, `markedDays`, `newGoal` (and by the way still the modal toggle flag 😉)
  
I suggest to refactor and simplify this structure and keep the goals ids close to the days for the DayPicker, e.g. 

```js
 const goals = [{ 
    goalId: "fbbe4105c94",
    days: [
    "2023-07-01T22:00:00.000Z",
    "2023-07-02T22:00:00.000Z",
    "2023-07-03T22:00:00.000Z",
    "2023-07-04T22:00:00.000Z",
   creationDate:  "27.7.2023",
   deadline: null,
   icon: "/icons/icons8-woman-walking-48.png",
   interval: "day",
   name: "Goals name",
   targetPerInterval: 3}]
```

This should be all you need to track your goals data. Each time you create a new goal, you add a new goal object to your goals array. Of course you have to change your `useEffect` hooks then to identify the goal by id from the goals array (filtering the array for `goalId`) and get the days from there for your picker, but this is imho a lot easier than the structure you are currently maintaining.
Hope that helps!

**Reply by AliaksandraNovikava** [reply]
Hey @actyvystom 

Thank you for the great suggestion on how to simplify my code structure👍 It made a huge difference in improving the readability and maintainability of my code🚀

But I still noticed that the DayPicker component sometimes seems to exhibit inconsistent and unexpected behavior, leading to changes in the number of selected days without apparent reason. The component might not behave predictably in certain scenarios, causing a varying number of selected days. I don't quite understand what the cause might be...

Here is my updated code: [new PR](https://github.com/AliaksandraNovikava/capstone-alex/pull/24)

Thank you again for your support🙏

**Comment by actyvystom** [comment, marked as answer]
Hi @AliaksandraNovikava, I prepared a stripped down working example page you can copy and set up as a demo page in your project (e.g. `my-goal/demo.js`):  
```jsx
import Modal from "@/components/Modal";
import { DayPicker } from "react-day-picker";
import { goals } from "@/lib/data";
import { parseISO } from "date-fns";
import { useState, useEffect } from "react";
import styled from "styled-components";
const StyledDescription = styled.p`
  font-size: 0.8em;
  font-style: italic;
  color: #8e8e93;
  margin: 20px 10px;
`;
// get some sample goals from the data
const myGoals = goals.slice(0, 2);
export default function SimplifiedGoals() {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [selectedGoal, setSelectedGoal] = useState();
  const [days, setDays] = useState([]);

  const footer =
    days && days.length > 0 ? (
      <StyledDescription>
        You selected <strong>{days.length}</strong> day(s) for{" "}
        <strong>{selectedGoal?.name}</strong>.
      </StyledDescription>
    ) : (
      <StyledDescription>Please pick one or more days.</StyledDescription>
    );
  useEffect(() => {
    const goalsData = JSON.parse(localStorage.getItem("goals"));
    if (goalsData) {
      const currentGoal = goalsData.find(
        (goal) => goal.id === selectedGoal?.id
      );
      if (currentGoal) {
        setDays(currentGoal.days.map((day) => parseISO(day)));
      } else {
        setDays([]);
      }
    }
  }, [selectedGoal]);

  useEffect(() => {
    if (!isModalOpen && selectedGoal) {
      const goalsRawData = JSON.parse(localStorage.getItem("goals"));
      if (goalsRawData) {
        const currentGoal = goalsRawData.find(
          (goal) => goal.id === selectedGoal.id
        );
        if (currentGoal) {
          currentGoal.days = days.map((day) => day.toISOString());
          localStorage.setItem("goals", JSON.stringify(goalsRawData));
        } else {
          goalsRawData.push({
            ...selectedGoal,
            days: days.map((day) => day.toISOString()),
          });
          localStorage.setItem("goals", JSON.stringify(goalsRawData));
        }
      } else {
        const newGoal = {
          ...selectedGoal,
          days: days.map((day) => day.toISOString()),
        };
        localStorage.setItem("goals", JSON.stringify([newGoal]));
      }
    }
  }, [isModalOpen, selectedGoal, days]);

  return (
    <>
      <ul>
        {myGoals.map((goal) => (
          <li
            key={goal.id}
            onClick={() => {
              setSelectedGoal(goal);
              setIsModalOpen(true);
            }}
          >
            {goal.name}
          </li>
        ))}
      </ul>
      {isModalOpen && (
        <Modal
          position="absolute"
          top="5%"
          closeModal={() => setIsModalOpen(!isModalOpen)}
        >
          <DayPicker
            mode="multiple"
            min={1}
            selected={days}
            onSelect={setDays}
            footer={footer}
          />
        </Modal>
      )}
    </>
  );
}

``` 
Please note that you should use `uid()` in your static goals data, this also simplifies a lot as you don't have to create new goal objects with  a new id then:

```js
import { uid } from "uid";
export const goals = [
  {
    id: uid(),
    name: "Walking",
    category: "Sport",
    description: "I want to take a walk after lunch or dinner",
    icon: "/icons/icons8-woman-walking-48.png",
  },
  {
    id: uid(),
    name: "Yoga",
    category: "Sport",
    description: "I want to do yoga",
    icon: "/icons/icons8-woman-in-lotus-position-48.png",
  },
...
```
❗ Keep in mind that using `uid()` creates new ids each time the application is (re-)started. So writing them to `localStorage` is only valid **once** for a running dev server session.  

:bulb: Take a look at this [VS Code extension - UUID Generator](https://marketplace.visualstudio.com/items?itemName=netcorext.uuid-generator) to create persisting ids for your static data.
